processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -407,35 +216,56 @@ public T get(int index)
   @Override
   public int indexOf(T value)
   {
-    return bufferIndexed.indexOf(value);
+    return indexOf(this, value);
+  }
+
+  private int indexOf(Indexed<T> indexed, T value)
```

Comment:
If this method is hot, JIT will figure out the class of this `indexed` and will inline the `get()` method.

However seems to me that this method is never hot.
 
Bigger problem with indexOf() is that it makes log(n) full deserializations of objects. It could be avoided if e. g. all Strings stored are ASCII, then we can compare bytes directly, without deserialization. I apply this optimization in the new dimension dictionary format, as part of #3882.